### ISSUE TITLE
Invalid model for QF_LRA formula (assertion violation at Debug.cc:85) 

### ISSUE THREAD

**Issue by rainoftime** [usi-verification-and-security/opensmt#92]
Related issue https://github.com/usi-verification-and-security/opensmt/issues/79 that was fixed in https://github.com/usi-verification-and-security/opensmt/pull/91

-----------

Hi, for the following instance
~~~~
(set-logic QF_LRA)
(declare-fun v6 () Bool)
(declare-fun v7 () Bool)
(declare-fun v13 () Bool)
(declare-fun v15 () Bool)
(assert (=> (=> v15 v6) (= v13 v7)))
(check-sat)
(check-sat)
(push 1)
(check-sat)
(pop 1)
(assert (= v15 (= v13 v7)))
(check-sat)
(push 1)
(check-sat)
(pop 1)
(assert (distinct true v15))
(push 1)
(check-sat)
(pop 1)
(push 1)
(check-sat)
(pop 1)
(push 1)
(check-sat)
(pop 1)
(push 1)
(check-sat)
(pop 1)
(check-sat)
~~~~

opensmt (commit a492c61689)  throws an assertion violation
~~~~
sat
sat
sat
sat
sat
unsat
unsat
unsat
unsat
unsatisfied clause: 3   7   ; (or v15 (or (not v15) v6))
unsatisfied clause: 11  4   ; (or (= v13 v7) (or v6 (not (= v13 v7))))
opensmt: /home/rainoftime/Work/tofuzz/opensmt/src/smtsolvers/Debug.cc:85: virtual void CoreSMTSolver::verifyModel(): Assertion `!failed' failed.
~~~~

**Comment by rainoftime** [comment]
Here is another test case
~~~~
(set-logic QF_LRA)
(declare-fun r14 () Real)
(push 1)
(check-sat)
(pop 1)
(declare-fun v12 () Bool)
(assert (or (<= 582095008.0 r14) v12))
(assert (and (=> v12 (<= 582095008.0 r14)) (not (<= 582095008.0 r14))))
(push 1)
(check-sat)
(push 1)
(push 1)
(push 1)
(push 1)
(push 1)
(check-sat)
(pop 1)
(pop 1)
(pop 1)
(check-sat)
(pop 1)
(check-sat)
(pop 1)
(pop 1)
(check-sat)
~~~~

**Comment by blishko** [comment]
Thanks, I see the problem.
You assert unsatisfiable formula, but you don't check-sat right away; you check-sat only after push.
Then, once you pop back to the original frame and check-sat there, the solver does not handle this situation correctly.

It hides the information that I relied upon in #91 

We will work on it soon, no need for additional test cases for this one!

**Comment by blishko** [comment]
Fixed by #141
